### PR TITLE
Refactor of planet difficulty in JSON

### PR DESCRIPTION
**Pull request by Steampunkery**
# Description
This PR replaces the `easyOnly` and `hardOnly` flags in `planetsConfig.json` with `difficultyLevels`. What this flag does is enable a planet to be created and placed in all modes. The previous system only allowed a planet to be in easy, medium, or hard mode alone. This is necessary because the game needs to be able to create at least one planet for each game mode. If the planets in the `core` module are disabled -- if you were to remove them from the assets directory for testing or game play reasons -- and you wanted to add a module that only had one planet, (again, for testing or game play) the game would crash because each planet can only be enabled for one mode. The `difficultyLevels` attribute is a list of all difficulties a planet should be spawned in, `"easy"`, `"medium"`, or `"hard"`. Thus allowing a planet to be in all modes, or any subset thereof, so the game doesn't crash.

# Testing
1. Checkout this branch
2. Get @BenjaminAmos' module, blackHole, `./groovyw module get blackHoles -remote BenjaminAmos`.
3. Remove the `grounds` directory from the `core` module, as well as the `planetsConfig.json` and `planetNamesConfig.json` (this step is to force the game to spawn only one type of planet -- black holes)
4. Launch the game, and see the crash.
5. Change the line that reads `"hardOnly" = true` in the `blackHole` module's `planetsConfig.json` to read:
```
"difficultyLevels": [
    "easy",
    "medium",
    "hard"
]
```
6. Relaunch the game and see that all planets are now black holes

### CODE REVIEW

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

File: engine/src/main/java/org/destinationsol/game/planet/PlanetConfig.java [hunk 1]
```diff
@@ -85,9 +84,17 @@ static PlanetConfig load(JsonValue rootNode, HullConfigManager hullConfigs, Game
...
+        String[] difficultyLevels = PlanetConfig.getDifficultyLevels(rootNode, moduleName);
         return new PlanetConfig(rootNode.name, minGrav, maxGrav, deco, groundEnemies, highOrbitEnemies, lowOrbitEnemies, cloudTextures,
-                planetTiles, stationConfig, skyConfig, rowCount, smoothLandscape, tradeConfig, hardOnly, easyOnly, moduleName);
+                planetTiles, stationConfig, skyConfig, rowCount, smoothLandscape, tradeConfig, difficultyLevels, moduleName);
+    }
+
+    static String[] getDifficultyLevels(JsonValue rootNode, String moduleName) {
+        try {
+            return rootNode.get("difficultyLevels").asStringArray();
+        } catch (NullPointerException | IllegalStateException e) {
```
Comment (Adrijaned) [inline engine/src/main/java/org/destinationsol/game/planet/PlanetConfig.java]: Please remove this try/catch - we now have JSON schemas!

File: engine/src/main/java/org/destinationsol/game/planet/PlanetConfigs.java [hunk 2]
```diff
@@ -52,12 +52,15 @@ public PlanetConfigs(HullConfigManager hullConfigs, GameColors cols, ItemManager
...
-                if (planetConfig.hardOnly) {
-                    hard.add(planetConfig);
-                } else if (planetConfig.easyOnly) {
-                    easy.add(planetConfig);
-                } else {
-                    medium.add(planetConfig);
+                for (String difficulty : planetConfig.difficultyLevels) {
+                    switch (difficulty) {
+                        case "easy":
+                            easy.add(planetConfig);
```
[the remark this answers is not in the document]
Comment (Adrijaned) [reply engine/src/main/java/org/destinationsol/game/planet/PlanetConfigs.java]:
```suggestion
                            easy.add(planetConfig);
                            break;
```

File: engine/src/main/java/org/destinationsol/game/planet/PlanetConfigs.java [hunk 3]
```diff
@@ -52,12 +52,15 @@ public PlanetConfigs(HullConfigManager hullConfigs, GameColors cols, ItemManager
...
+                        case "medium":
+                            medium.add(planetConfig);
```
[the remark this answers is not in the document]
Comment (Adrijaned) [reply engine/src/main/java/org/destinationsol/game/planet/PlanetConfigs.java]:
```suggestion
                            medium.add(planetConfig);
                            break;
```

File: engine/src/main/java/org/destinationsol/game/planet/PlanetConfigs.java [hunk 4]
```diff
@@ -52,12 +52,15 @@ public PlanetConfigs(HullConfigManager hullConfigs, GameColors cols, ItemManager
...
+                        case "hard":
+                            hard.add(planetConfig);
```
[the remark this answers is not in the document]
Comment (Adrijaned) [reply engine/src/main/java/org/destinationsol/game/planet/PlanetConfigs.java]:
```suggestion
                            hard.add(planetConfig);
                            break;
```

File: modules/core/assets/configs/planetsConfig.json [hunk 5]
```diff
@@ -1,7 +1,9 @@
 {
     "rocky": {
-        "easyOnly": true,
-        "minGrav": 0.4,
+		"difficultyLevels": [
```
Comment (Adrijaned) [inline modules/core/assets/configs/planetsConfig.json]: Please, don't use tabs. Ever.

File: modules/core/assets/configs/planetsConfig.json [hunk 6]
```diff
@@ -85,6 +87,9 @@
     },
 
     "desert": {
+		"difficultyLevels": [
```
Comment (Adrijaned) [inline modules/core/assets/configs/planetsConfig.json]: Tabs

File: modules/core/assets/configs/planetsConfig.json [hunk 7]
```diff
@@ -179,8 +184,10 @@
 
 
     "frozen": {
-        "hardOnly": true,
-        "minGrav": 0.8,
+		"difficultyLevels": [
```
Comment (Adrijaned) [inline modules/core/assets/configs/planetsConfig.json]: Tabs

**Comment by NicholasBatesNZ** [comment]
Hey @Steampunkery are you planning on finishing this up? Looks pretty close :)

**Comment by NicholasBatesNZ** [comment]
Closing as seemingly abandoned and I don't quite see the point. Why would a module author want a planet to be hard.. and easy.. at the same time??

**Pull request closed without merging**